### ISSUE TITLE
SetColStyle not honoring given style

### ISSUE THREAD

**Issue by SteveTalt** [360EntSecGroup-Skylar/excelize#467]
**Description**
SetColStyle seems to be corrupting spreadsheets.    It doesn't seem to matter what data is in the column or which column or sheet you use.

I created this simple code to re-produce it:

```
package main

import (
	"github.com/excelize"
)

// define globals

var xlfile = "testlocal.xlsx"
var xlsheetname = "test"

// main
func main() {
	var (
		file *excelize.File
		err  error
	)
	// open the file
	file, err = excelize.OpenFile(xlfile)
	errorhandler(err)

	err = file.SetColStyle(xlsheetname, "C", 7)
	errorhandler(err)
	file.Save()

}

func errorhandler(e error) {
	if e != nil {
		panic(e)
	}
}
```

**Steps to reproduce the issue:**
1.   Create a spreadsheet named localtest.xlsx and a sheet named test
2.   run this test code against it
3.   open the spreadsheet and see it needs repair

**Describe the results you received:**
![image](https://user-images.githubusercontent.com/12849870/62726322-d97f2980-b9cb-11e9-81fe-ab3a47d6daed.png)
Repair Log:
`
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<recoveryLog xmlns="http://schemas.openxmlformats.org/spreadsheetml/2006/main"><logFileName>Repair Result to testlocal5.xml</logFileName><summary>Errors were detected in file '/Users/stalt/Documents/go/src/setcoltest/testlocal.xlsx'</summary><repairedRecords summary="Following is a list of repairs:"><repairedRecord>Repaired Records: Column information from /xl/worksheets/sheet1.xml part</repairedRecord></repairedRecords></recoveryLog>

`

**Describe the results you expected:**
I expect the column to be formated with 

**Output of `go version`:**

```text
$ go version
go version go1.10 darwin/amd64
```

**Excelize version or commit ID:**

```text
commit id: 448f552
```

**Environment details (OS, Microsoft Excel™ version, physical, etc.):**
macOS Mojave 10.14.5
Excel for Mac 16.27 (19071500)

**Comment by SteveTalt** [comment]
Weirdly, if i do this:
	
`style, err = file.NewStyle(`{"number_format": 7}`)
err = file.SetColStyle(xlsheetname, "C", style)`

the value of style is 2?? and it formatting to 2 style and doesn't corrupt it.   It may be that format 7 doesn't exist.

**Comment by SteveTalt** [comment]
I'm expecting 7 to be: 
![image](https://user-images.githubusercontent.com/12849870/62730356-a4c3a000-b9d4-11e9-9054-9587d10cad9c.png)

based on [this](https://xuri.me/excelize/en/style.html#number_format)

**Comment by SteveTalt** [comment]
Created a custom style....doesn't break it...but also doesn't change the style:

```
style, err = f.NewStyle(`{"custom_number_format": "($#,##0.00_);($#,##0.00)"}`)
err = file.SetColStyle(xlsheetname, "C", style)
```

**Comment by SteveTalt** [comment]
For reference, it works fine using SetCellStyle with the same custom format.   Just not SetColStyle.

**Comment by mlh758** [comment]
You're getting the error with `SetColStyle` because [the function](https://godoc.org/github.com/360EntSecGroup-Skylar/excelize#File.SetColStyle) expects the ID of a custom style already added to the sheet. If you just put a number in before adding a style with that ID then the generated relationship is incorrect. The way you have it in your later comments with `NewStyle` is the correct way to use it.

**Comment by SteveTalt** [comment]
Understood.   BUT:

1)  using new style with SetColStyle still does not work....it doesn't corrupt the file...but also doesn't format the column...it basically does nothing.     It works with SetCellStyle...but not SetColStyle
2) I wasn't trying to use a custom style....i was trying to use an number_format that already existed.   If it requires a custom_number_format it should say that in the documentation...it doesn't.    There is a whole section in the documentation that is a table of number_format and styleIDs which thus makes it very unclear that you'd have to create your own style to use this function.
3) even if the function can't work without newstyle....it shouldn't corrupt the file as a result and should error out instead....no error is thrown when this happens.

I'll update my sample code with new style and setcolstyle and repost in a bit.

**Comment by mlh758** [comment]
> I wasn't trying to use a custom style....i was trying to use an number_format that already existed.

Even so, you would have to use `NewStyle` to create a style in the document to apply to your cell. Anything that sets a style in this library takes the style ID for a style that already exists in the workbook. `7` isn't an existing style ID, it's one of the built in number formats that you can apply to a style you create before calling set style.

As for creating the style, they're touchy since you have to get the JSON just right, I look forward to something like #470 simplifying this whole interaction. Are you checking the error from `NewStyle`?

**Comment by SteveTalt** [comment]
Yes, I am checking for error in new style....and I know this new style works because in my real code when I apply it via setcellstyle it works.   Here is an updated version:

```
package main

import (
	"github.com/excelize"
)

// define globals

var xlfile = "testlocal.xlsx"
var xlsheetname = "test"

// main
func main() {
	var (
		file  *excelize.File
		err   error
		style int
	)
	// open the file
	file, err = excelize.OpenFile(xlfile)
	errorhandler(err)

	// set the cost column to be the number format I want
	style, err = file.NewStyle(`{"custom_number_format": "#,##0.00;(#,##0.00)"}`)

	errorhandler(err)
	err = file.SetColStyle(xlsheetname, "C", style)
	errorhandler(err)
	file.Save()

}

func errorhandler(e error) {
	if e != nil {
		panic(e)
	}
}
```

**Comment by mlh758** [comment]
Could you change the title of this issue to something like "SetColStyle not honoring given style" or something? The repair thing ended up not being an issue and when we figure this out it will help people search for it later if they have a similar problem.

**Comment by SteveTalt** [comment]
Interestingly, if I change the style to this:

`style, err = file.NewStyle(`{"fill":{"type":"gradient","color":["#FFFFFF","#E0EBF5"],"shading":1}}`)`

it does the fill....but starts  right below where my data is:

![image](https://user-images.githubusercontent.com/12849870/62794495-1b21da00-ba89-11e9-9ef2-1215a9714fec.png)

**Comment by mlh758** [comment]
I think we're just missing 7 in the built-in table:
https://github.com/360EntSecGroup-Skylar/excelize/blob/master/styles.go#L29

**Comment by mlh758** [comment]
You can use `unzip file.xlsx` to see what xml was generated. That can help troubleshoot what styles are created and applied.

**Comment by mlh758** [comment]
If I apply a style to a column in Excel I get both a column level style set:
```
<cols>
    <col min="1" max="1" width="10.83203125" style="1"/>
</cols>
```

and each row:

```
<row r="1" spans="1:1" x14ac:dyDescent="0.2">
    <c r="A1" s="1">
        <v>123</v>
    </c>
</row>

**Comment by SteveTalt** [comment]
Here is the styles XML file:

[styles.xml.zip](https://github.com/360EntSecGroup-Skylar/excelize/files/3487802/styles.xml.zip)

This would have the gradient fill since that was the last style i saved to that column.    I see it defined in there.

Here is the sheet xml:
[sheet1.xml.zip](https://github.com/360EntSecGroup-Skylar/excelize/files/3487852/sheet1.xml.zip)

I see the col style and row style assigned (but they aren't the same).   Looks like column style is 3 and row style is 2 if i'm reading that right.

column: 

 ```
 <cols>
    <col collapsed="false" hidden="false" max="3" min="3" style="3" width="9">
    </col>
  </cols>
```

Example Row:

```
    <row r="5" spans="1:3">
      <c r="A5" t="s">
        <v>6</v>
      </c>
      <c r="B5" t="s">
        <v>6</v>
      </c>
      <c r="C5" s="2">
        <v>1234.5</v>
      </c>
    </row>
```

**Comment by SteveTalt** [comment]
Style 2 has a different FillID.   Style 3 looks right.

So it looks like for the rows with data in them....the row style is over riding the column style.

**Comment by mlh758** [comment]
It looks like we're adding borders into styles.xml every time `NewStyle` is called. That probably shouldn't be happening. Unrelated to this though.

**Comment by mlh758** [comment]
The styles and sheet files you attached, those correspond to the file you posted the screenshot of with the gradient?

**Comment by SteveTalt** [comment]
Yes....correct.   It's the same sheet I was using before with the custom_number_format, but the last that screenshot of the gradient is the current state.     

All I did to create it was replace:

```style, err = file.NewStyle(`{"custom_number_format": "#,##0.00;(#,##0.00)"}`)```

with 

```
style, err = file.NewStyle({"fill":{"type":"gradient","color":["#FFFFFF","#E0EBF5"],"shading":1}})
```

and re-ran it.

**Comment by emdioh** [comment]
I think that the issue is in SetColStyle implementation: if there are no column ranges defined in the worksheet the function ultimately does nothing.
I am going to send a PR to fix this.

**Comment by xuri** [comment]
This issue has been fixed by commit [`4f0d676`](https://github.com/360EntSecGroup-Skylar/excelize/commit/4f0d676eb765472d1fe7a29cacd165b982785bd2), and this patch will be released in the next version.
